### ISSUE TITLE
add "enabled=True" to DistributedDataParallel.no_sync()

### ISSUE THREAD

**Issue by avihu111** [pytorch/pytorch#143721]
### 🚀 The feature, motivation and pitch

Training a model with DDP and gradient accumulation is quite common.
To avoid unnecessary sync, the no_sync() operation is used.

Providing an `enabled=True` argument is already done in pytorch, and is very useful in pytorch in `torch.amp.autocast` and `torch.amp.GradScaler`.

```
if (step % grad_accum_steps + 1) == 0:
    # forward+ backward code
    loss = ddp_model(inputs)
    (loss / grad_accum_steps).backward()
else:
    with ddp_model.no_sync():
         # forward + backward code
         loss = ddp_model(inputs)
        (loss / grad_accum_steps).backward()
```
using the `enabled` argument this can be simplified, preventing bug-prone code duplications:
```
with ddp_model.no_sync(enabled=(step % grad_accum_steps + 1) != 0):
    loss = ddp_model(inputs)
    (loss / grad_accum_steps).backward()
```

The implementation doesn't seem hard, and it will be back-compatible.

### Additional context

DDP with grad accum:
https://discuss.pytorch.org/t/gradient-accumulation-with-ddp-no-sync-interface/169593/3

Current no_sync implementation:
https://github.com/pytorch/pytorch/blob/main/torch/nn/parallel/distributed.py#L1420

torch.amp.autocast enabled=True API:
https://github.com/pytorch/pytorch/blob/09c950cc872dfcee453307db47fa10553c3f5616/torch/amp/autocast_mode.py#L222

**Comment by LunNova** [comment]
Examples of not having enabled making code elsewhere less elegant:

https://github.com/karpathy/nanoGPT/blob/93a43d9a5c22450bbf06e78da2cb6eeef084b717/train.py#L293-L301

```py
        if ddp:
            # in DDP training we only need to sync gradients at the last micro step.
            # the official way to do this is with model.no_sync() context manager, but
            # I really dislike that this bloats the code and forces us to repeat code
            # looking at the source of that context manager, it just toggles this variable
            model.require_backward_grad_sync = (micro_step == gradient_accumulation_steps - 1)
        with ctx:
            logits, loss = model(X, Y)
            loss = loss / gradient_accumulation_steps # scale the loss to account for gradient accumulation
```

https://github.com/KellerJordan/modded-nanogpt/blob/f75897d7424458f169c95050ea7fbbe40ada20f1/train_gpt2.py#L589-L594

```py
        with contextlib.ExitStack() as stack:
            if i < train_accumulation_steps: # there's no need to sync gradients every accumulation step
                stack.enter_context(model.no_sync())
            if step >= 5:
                stack.enter_context(torch.compiler.set_stance(skip_guard_eval_unsafe=True))
            model(inputs_train, targets_train, sliding_window_num_blocks).backward()
```

**Comment by yf225** [comment]
IMO it makes sense to have this feature. Would you like to open a PR for this?

**Comment by avihu111** [comment]
Awesome, I'll work on the PR and update 💯

**Comment by avihu111** [comment]
Opened!
https://github.com/pytorch/pytorch/pull/143939

**Comment by upskyy** [comment]
It’s also possible to implement it this way. I’m leaving this here in case it might be helpful to someone in the future.

```python
from contextlib import nullcontext


sync_context = nullcontext()
if (step % grad_accum_steps + 1) != 0: # the option you want
    sync_context = ddp_model.no_sync()

with sync_context:
    loss = ddp_model(inputs)
    (loss / grad_accum_steps).backward()

```
